File: python_algo/divide.py
from additional import *

REVERSE_ROTATE = -1
ROTATE = 1
NO_DIR = 3


def mark_greater_than(current_index, stack_arr):
    elem = stack_arr[current_index]
    stack_arr[current_index].keep = True
    ind = current_index + 1
    if ind > len(stack_arr) - 1:
        ind = 0
    while stack_arr[ind] != elem:
        if stack_arr[ind].order > stack_arr[current_index].order:
            stack_arr[ind].keep = True
            current_index = ind
        else:
            stack_arr[ind].keep = False
        ind += 1
        if ind > len(stack_arr) - 1:
            ind = 0


def find_nearest_false(to_look):
    nearest = len(to_look)
    rot_dir = NO_DIR
    for i in range(int(len(to_look) / 2)):
        if to_look[i].keep == False and i < nearest:
            nearest = i
            rot_dir = REVERSE_ROTATE
    for i in range(int(len(to_look) / 2), len(to_look)):
        if to_look[i].keep == False and len(to_look) - i < nearest:
            nearest = len(to_look) - i
            rot_dir = ROTATE
    return nearest, rot_dir


def mark_(arr_to_mark):
    last = len(arr_to_mark)
    for elem in arr_to_mark:
        if elem.order == last \
                or elem.order + 1 == last \
                or elem.order + 2 == last:
            elem.keep = True

def divide_(to_div_arr, stack_A, stack_B):
    mark_(stack_A.arr)
    while True:
        if not to_div_arr[0].keep:
            push_('b', stack_A, stack_B)
        else:
            nearest, rot_dir = find_nearest_false(to_div_arr)
            if rot_dir == ROTATE:
                while nearest > 0:
                    rotate_('a', stack_A, stack_B)
                    nearest -= 1
            elif rot_dir == REVERSE_ROTATE:
                while nearest > 0:
                    reverse_rotate_('a', stack_A, stack_B)
                    nearest -= 1

        if len(to_div_arr) <= 3:
            while True:
                # clear_keep(to_div_arr)
                for i in range(len(to_div_arr)):
                    mark_greater_than(i, to_div_arr)
                    if not no_keep_elem_count(to_div_arr):
                        break
                if not no_keep_elem_count(to_div_arr):
                    break
                else:
                    swap_('a', stack_A, stack_B, True)
            break
        elif not no_keep_elem_count(stack_A.arr):
            break

File: push_swap/divide/divide.c
#include "push_swap.h"

inline static void	div_rot(int steps, int rot_dir,
							t_list *stack_a, t_list *stack_b)
{
	if (rot_dir == ROTATE)
	{
		while (steps--)
			rotate('a', stack_a, stack_b);
	}
	else if (rot_dir == REVERSE_ROTATE)
	{
		while (steps--)
			reverse_rotate('a', stack_a, stack_b);
	}
}

static uint8_t	sort_small(t_list *stack_a)
{
	while (TRUE)
	{
		stack_a->cur = stack_a->head;
		while (TRUE)
		{
			mark_greater_than(*stack_a);
			stack_a->cur = stack_a->cur->next;
			if (stack_a->cur->prev == stack_a->end
				|| !not_keep_elem_count(*stack_a))
				break ;
		}
		if (!not_keep_elem_count(*stack_a))
			break ;
		else
			swap(stack_a);
	}
	return (1);
}

static void	find_nearest_false(t_list to_look, int *steps, int *rot_dir)
{
	*steps = ft_lstsize(to_look);
	*rot_dir = NO_DIR;
	to_look.cur = to_look.head;
	while (TRUE)
	{
		if (ft_lstindex(to_look) < ft_lstsize(to_look) / 2
			&& !((t_content *)to_look.cur->content)->keep
			&& ft_lstindex(to_look) < *steps)
		{
			*steps = ft_lstindex(to_look);
			*rot_dir = ROTATE;
		}
		else if (ft_lstindex(to_look) >= ft_lstsize(to_look) / 2
			&& !((t_content *)to_look.cur->content)->keep
			&& ft_lstsize(to_look) - ft_lstindex(to_look) < *steps)
		{
			*steps = ft_lstsize(to_look) - ft_lstindex(to_look);
			*rot_dir = REVERSE_ROTATE;
		}
		to_look.cur = to_look.cur->next;
		if (to_look.cur->prev == to_look.end)
			break ;
	}
}

void	divide(t_list *stack_a, t_list *stack_b)
{
	int	steps;
	int	rot_dir;

	if (is_sort(*stack_a))
		return ;
	ft_lstiter(*stack_a, clear_keep);
	ft_lstiter(*stack_a, mark);
	stack_a->cur = stack_a->head;
	while (TRUE)
	{
		if ((ft_lstsize(*stack_a) <= 3 && sort_small(stack_a))
			|| !not_keep_elem_count(*stack_a))
			break ;
		else if (!((t_content *)stack_a->head->content)->keep)
			push('b', stack_a, stack_b);
		else
		{
			find_nearest_false(*stack_a, &steps, &rot_dir);
			div_rot(steps, rot_dir, stack_a, stack_b);
		}
	}
}
